smack-extensions/src/main/java/org/jivesoftware/smackx/avatar/UserAvatarManager.java

```diff
@@ -0,0 +1,344 @@
+
+    static {
+        XMPPConnectionRegistry.addConnectionCreationListener(new ConnectionCreationListener() {
+            @Override
+            public void connectionCreated(XMPPConnection connection) {
+                getInstanceFor(connection);
+            }
+        });
+    }
+
+    private static final Map<XMPPConnection, UserAvatarManager> INSTANCES = new WeakHashMap<>();
+
+    /**
+     * Get the singleton instance of UserAvatarManager.
+     *
+     * @param connection
+     * @return the instance of UserAvatarManager
+     */
+    public static synchronized UserAvatarManager getInstanceFor(XMPPConnection connection) {
+        UserAvatarManager userAvatarManager = INSTANCES.get(connection);
+
+        if (userAvatarManager == null) {
+            userAvatarManager = new UserAvatarManager(connection);
+            INSTANCES.put(connection, userAvatarManager);
+        }
+
+        return userAvatarManager;
+    }
+
+    private UserAvatarManager(XMPPConnection connection) {
+        super(connection);
+    }
+
+    /**
+     * Returns true if User Avatar is supported by the server.
+     * 
+     * @return true if User Avatar is supported by the server.
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     */
+    public boolean isSupportedByServer()
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+        return findNodeItem(DATA_NAMESPACE) && findNodeItem(METADATA_NAMESPACE);
+    }
+
+    private boolean findNodeItem(String nodeName)
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+        ServiceDiscoveryManager serviceDiscoveryManager = ServiceDiscoveryManager.getInstanceFor(connection());
+        List<Item> items = serviceDiscoveryManager.discoverItems(connection().getUser().asBareJid(), nodeName)
+                .getItems();
+        return items != null && items.size() > 0;
+    }
+
+    /**
+     * Get the data node.
+     * 
+     * @return the data node
+     * @throws NoResponseException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     * @throws XMPPErrorException
+     */
+    public LeafNode getDataNode()
+            throws NoResponseException, NotConnectedException, InterruptedException, XMPPErrorException {
+        return getNode(DATA_NAMESPACE);
+    }
+
+    /**
+     * Get the metadata node.
+     * 
+     * @return the metadata node
+     * @throws NoResponseException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     * @throws XMPPErrorException
+     */
+    public LeafNode getMetadataNode()
+            throws NoResponseException, NotConnectedException, InterruptedException, XMPPErrorException {
+        return getNode(METADATA_NAMESPACE);
+    }
+
+    private LeafNode getNode(String nodeName)
```

Comment:
That is why it's a good idea to have `PubSubManager.getOrCreateNode(String)`, you are duplicating code that you already wrote for Geo Location here.